### PR TITLE
MeasureImageSkeleton (resolves #2689)

### LINKED ISSUE #2689: MeasureImageSkeleton and MeasureObjectSkeleton

### PR DESCRIPTION
**Pull request by 0x00b1**
screenshot

### CODE REVIEW

**Comment by 0x00b1** [comment]
cc: @bethac07

File: cellprofiler/modules/measureimageskeleton.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
+import cellprofiler.module
+import cellprofiler.setting
+import numpy
+import scipy.ndimage
+import skimage.measure
+import skimage.segmentation
+import skimage.util
```
Comment (mcquin) [inline cellprofiler/modules/measureimageskeleton.py]: Imports can be optimized.

File: cellprofiler/modules/measureimageskeleton.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+
+def _neighbors(image):
```
Comment (mcquin) [inline cellprofiler/modules/measureimageskeleton.py]: Can you please add a comment explaining this procedure? A link to an article would be fine, too.

File: cellprofiler/modules/measureimageskeleton.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+def branches(image):
+    return _neighbors(image) > 2
+
+
+def endpoints(image):
+    return _neighbors(image) == 1
+
+
+class MeasureImageSkeleton(cellprofiler.module.Module):
```
Comment (mcquin) [inline cellprofiler/modules/measureimageskeleton.py]: Documentation.

File: cellprofiler/modules/measureimageskeleton.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    category = "Measurement"
+
+    module_name = "MeasureImageSkeleton"
+
+    variable_revision_number = 1
+
+    def create_settings(self):
+        self.skeleton_name = cellprofiler.setting.ImageNameSubscriber(
```
Comment (mcquin) [inline cellprofiler/modules/measureimageskeleton.py]: Documentation.

File: cellprofiler/modules/measureimageskeleton.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+import skimage.segmentation
+import skimage.util
+
+
+def _neighbors(image):
+    padding = skimage.util.pad(image, 1, "constant")
+
+    mask = padding > 0
+
+    padding = padding.astype(numpy.float)
```
Comment (mcquin) [inline cellprofiler/modules/measureimageskeleton.py]: Do you want to convert to float before padding by 1? As I understand, `image` is of Boolean type.

File: cellprofiler/modules/measureimageskeleton.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+        measurements = workspace.measurements
+
+        measurement_name = self.skeleton_name.value
+
+        statistics = []
+
+        name = "Skeleton_Branches_{}".format(measurement_name)
+
+        value = numpy.count_nonzero(skimage.measure.label(branches(data)))
```
Comment (mcquin) [inline cellprofiler/modules/measureimageskeleton.py]: Is `skimage.measure.label` necessary? `numpy.count_nonzero` counts `True` as nonzero.

File: cellprofiler/modules/measureimageskeleton.py (hunk 3 above)
Comment (mcquin) [inline cellprofiler/modules/measureimageskeleton.py]: Additionally, you're going to need to implement a few more methods. According to the [module documentation](https://github.com/CellProfiler/CellProfiler/blob/8472a4b0b04879169b0a90b67a9e216658d53d5d/cellprofiler/module.py#L35-L40), you need to implement:

    get_categories - The category of measurement produced, for instance AreaShape
    get_measurements - The measurements produced by a category
    get_measurement_images - The images measured for a particular measurement
    get_measurement_scales - the scale at which a measurement was taken
    get_measurement_columns - the measurements stored in the database

File: cellprofiler/modules/measureimageskeleton.py (hunk 1 above)
Comment (0x00b1) [inline cellprofiler/modules/measureimageskeleton.py]: How so?

Comment (mcquin) [inline cellprofiler/modules/measureimageskeleton.py]: See the second bullet point under the [imports section of PEP08](https://www.python.org/dev/peps/pep-0008/#imports). PyCharm can do this for you if you navigate to Code > Optimize Imports.

File: cellprofiler/modules/measureimageskeleton.py (hunk 5 above)
Comment (0x00b1) [inline cellprofiler/modules/measureimageskeleton.py]: I use `astype(numpy.float)` to ensure `scipy.ndimage.uniform_filter` returns a `numpy.float`.

File: cellprofiler/modules/measureimageskeleton.py (hunk 2 above)
Comment (0x00b1) [inline cellprofiler/modules/measureimageskeleton.py]: ✅

File: cellprofiler/modules/measureimageskeleton.py (hunk 1 above)
Comment (0x00b1) [inline cellprofiler/modules/measureimageskeleton.py]: TIL!

Comment (0x00b1) [inline cellprofiler/modules/measureimageskeleton.py]: ✅

File: cellprofiler/modules/measureimageskeleton.py [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+                [0, 3, 0]
+            ]
+
+    :type image: numpy.ndarray
+
+    :param image: A two-or-three dimensional image
+
+    :return: neighbor pixels for each pixel of an image
+
+    """
```
Comment (mcquin) [inline cellprofiler/modules/measureimageskeleton.py]: So good!!! Thanks!

File: cellprofiler/modules/measureimageskeleton.py (hunk 3 above)
Comment (mcquin) [inline cellprofiler/modules/measureimageskeleton.py]: Done!

**Pull request merged**